### PR TITLE
Track alignment in ir

### PR DESCRIPTION
**Pull request by abadams**
This branch tracks alignment of loads or stores as annotations inside the load and store nodes themselves. The values are populated by the simplifier (this branch is built on #3402). This removes the need for other visitors to track alignment info if they're just using it for load and store alignment. I removed alignment tracking from hexagon, but there were a few places where I wasn't totally sure what to do. I marked those with TODO(dsharlet).

This diff will be easier to view once #3402 is in.

### CODE REVIEW

**Comment by abadams** [comment]
Hrm, based on the failures I think this is thinking something is more aligned than it actually is. Needs some diagnosing.

File: src/AlignLoads.cpp [hunk 1]
```diff
@@ -36,7 +36,9 @@ class AlignLoads : public IRMutator2 {
     Expr make_load(const Load *load, Expr index) {
         internal_assert(is_one(load->predicate)) << "Load should not be predicated.\n";
         return mutate(Load::make(load->type.with_lanes(index.type().lanes()), load->name,
-                                 index, load->image, load->param, const_true(index.type().lanes())));
+                                 index, load->image, load->param,
+                                 const_true(index.type().lanes()),
+                                 ModulusRemainder() /* TODO(dsharlet) */));
```
[the remark this answers is not in the document]
Comment (abadams) [reply src/AlignLoads.cpp]: There's a simplification pass that runs after align loads, which will reinfer alignment anyway, so I'll just skip it here.

File: src/LoopCarry.cpp [hunk 2]
```diff
@@ -395,22 +395,22 @@ class LoopCarryOverLoop : public IRMutator2 {
                 const Load *orig_load = loads[c[i]][0];
                 Expr scratch_idx = scratch_index(i, orig_load->type);
                 Expr load_from_scratch = Load::make(orig_load->type, scratch, scratch_idx,
-                                                    Buffer<>(), Parameter(), const_true(orig_load->type.lanes()));
+                                                    Buffer<>(), Parameter(), const_true(orig_load->type.lanes()), ModulusRemainder());
```
[the remark this answers is not in the document]
Comment (abadams) [reply src/LoopCarry.cpp]: Yeah, the simplifier runs after this, but also the loads from scratch are loads at constant indices, so the alignment field isn't necessary for llvm to get the correct alignment.

**Comment by abadams** [comment]
I think there's a small problem with this approach. We're relying on the simplifier running on the offloaded hexagon code to infer alignment after hexagon optimization passes, but at that point, all the asserts have been lost. We no longer pass in an alignment scope for int32 arguments because we can't easily populate that scope in a way that exploits the alignment constraints.

So I guess we just can't rely on the simplifier fixing things up for the hexagon passes. Nuts.

File: src/CodeGen_Internal.cpp [hunk 3]
```diff
@@ -308,7 +308,7 @@ class UnpredicateLoadsStores : public IRMutator2 {
                 Expr idx_i = Shuffle::make({index_var}, {i});
                 Expr pred_i = Shuffle::make({predicate_var}, {i});
                 Expr unpredicated_load = Load::make(op->type.element_of(), op->name, idx_i, op->image, op->param,
-                                                    const_true());
+                                                    const_true(), ModulusRemainder());
```
[the remark this answers is not in the document]
Comment (abadams) [reply src/CodeGen_Internal.cpp]: I don't think so - the indices are shuffles, not additions to a base.

**Comment by abadams** [comment]
Yes, but my implementation when I wrote that comment did indeed lose information, and rely on a late simplification pass to recover it. I have pushed a commit that hopefully doesn't lose alignment information this time, though it will need hexagon testing.

File: src/AlignLoads.cpp [hunk 4]
```diff
@@ -124,42 +129,20 @@ class AlignLoads : public IRMutator2 {
             // address. Rewrite this is an aligned load of two
             // native vectors, followed by a shuffle.
             Expr aligned_base = simplify(ramp->base - (int)aligned_offset);
-            Expr aligned_load = make_load(op, Ramp::make(aligned_base, 1, lanes*2));
+            ModulusRemainder alignment = op->alignment - (int)aligned_offset;
+            Expr aligned_load = make_load(op, Ramp::make(aligned_base, 1, lanes*2), alignment);
```
Comment (dsharletg) [inline src/AlignLoads.cpp]: It would be awesome to assert that alignment.remainder is 0 here.

**Comment by dsharletg** [comment]
I tested this on my Hexagon code and it all looks correct. I'm seeing some aligned loads/stores, but not everything that should be aligned is yet. I think it makes sense to get this PR producing correct results and merged and then I can more easily debug the finer points in follow-up PRs.

**Comment by abadams** [comment]
> ahhh -- is this why we were still getting intermittent failures on the buildbots? (ie maybe one of them is AVX but not AVX2?)

I believe so, yes.

**Pull request merged**